### ISSUE TITLE
Better support for :not(%placeholder) where %placeholder hasn't been extended

### ISSUE THREAD

**Issue by fabiospampinato** [sass/sass#2007]
At the moment the following SCSS:

```
%foo {}

.bar:not(%foo) {
  prop: val;
}
```

will compile to nothing, in my opinion this is not the expected behaviour, it should get compiled to:

```
.bar {
  prop: val;
}
```

the fact that `%foo` has never been extended, in the case where it's inside a `:not()`, doesn't mean that the selector will never match anything.

**Comment by cimmanon** [comment]
This does not seem to be consistent with how placeholder behaves under any other circumstance.  I do not see why anyone else would believe that a selector that contains a placeholder selector would exist in the compiled output when it hasn't been extended.

```
.foo%bar {
    prop: val;
}
```

**Comment by fabiospampinato** [comment]
I'm only talking about when inside a `:not` pseudo-class, not in general.

If fact in the example that you provided nothing should change at all. `.foo%bar` matches an element that has the class `foo` and also extends `%bar`, therefor if `%bar` is not extended then the selector won't match anything and shouldn't compile.

But in the example I provided an element matches the selector if has the class `bar` but is not an element that extends `%foo`, and if no element actually extends `%foo` than it means "every `.bar` element that is not nothing", see the double negation, basically just "every `.bar`".

I don't see any inconsistency.

**Comment by chriseppstein** [comment]
Sass is working correctly here. `.bar:not(%foo)` means "All the elements with class `bar` that are not also `%foo`". There is no `%foo`, so there's no need for the `not()` pseudoclass. To verify this logic, consider `.bar:not(span)` in a world where you can verify without a doubt that there is no `span` element with class `.bar`; the `not` is pointless in that situation and can be safely removed. Similarly, Sass knows there are no elements matching `%foo` so it optimizes your output for you.

**Comment by fabiospampinato** [comment]
There are at least 2 fundamental flaws in your reasoning:
1. In the example case I provided we know for sure that the pseudo-class is not needed, but what if we couldn't know? Just to provide a real life scenario that may actually happen: I'm writing a front-end framework and the user can choose to partially build it, basically avoiding the compilation of some files, and it's a possibility that said user don't want the default `icon` module, containing the `%icon` placeholder; now let's say in another module we have to apply some rules to children that are not `.baz` nor `%icon`, I would then write `.module > :not(.baz):not(%icon)` as everybody else would do. As you can see given the bug I reported this selector will never compile, and even in the case we have a `.baz` children the style won't be applied. Sure, I can double the amount of selectors and associated properties to workaround this bug, but it would only help having a less maintainable code.
2. The fact that a selector is pointless doesn't mean that it shouldn't work, as long as it's valid. I totally expect to be allowed to write `body:not(html)` and get it compiled. No doubt it would be silly, but it must work. Am I right?

Plus I see no reasons why fixing this bad behaviour implies the inability to optimize the cases where placeholders don't get extended. Will the compiler require some rewrites? Sure, but this is never a good reason not to fix a bug.

**Comment by chriseppstein** [comment]
> but what if we couldn't know?

We _can know_ because placeholders are a sass stylesheet concept and the compiler has examined the entire stylesheet and all imported stylesheets and knows definitively that there are not elements matching `%bar` because it was not extended by a concrete selector.

> The fact that a selector is pointless doesn't mean that it shouldn't work, as long as it's valid.

Agreed. you're making my point here.  `body:not(html)` is always equivalent to `body`; removing the not has no functional difference. you're requesting that we remove the `body` selector and the associated rules which _would be_ a functional difference that breaks stylesheets.

> Will the compiler require some rewrites? Sure, but this is never a good reason not to fix a bug.

I did not make the argument that the change would be hard, I made the argument that the software is working as expected and that you have a faulty understanding of it.

**Comment by chriseppstein** [comment]
@fabiospampinato ah I think I see the issue here. we're in violent agreement. I thought you were requesting that we don't output anything instead of just omitting the `:not()`.

TL;DR we both fail at reading comprehension.

**Comment by fabiospampinato** [comment]
My bad for haven't been clearer.

> I thought you were requesting that we don't output anything instead of just omitting the `:not()`.

Outputting nothing is exactly what currently happens, as said in my first post.

I'm glad the we finally come to a mutual understanding.
